### PR TITLE
⭐   Adding support to get the MongoDB and PostgreSQL cosmodb Types (Azure)

### PR DESCRIPTION
**Pull request by HRouhani**
Screenshot from 2024-07-29 18-49-59

Based on the current capabilities of the Azure SDK for Go, it appears that the kind, tags, and specially **properties** attributes are not directly available for Cosmodb type MongoDB and PostgreSQL server groups through the SDK.

### CODE REVIEW

File: providers/azure/resources/cosmosdb.go [hunk 1]
```diff
@@ -75,10 +96,92 @@ func (a *mqlAzureSubscriptionCosmosDbService) accounts() ([]interface{}, error)
...
+		if err != nil {
+			return err
+		}
+		for _, account := range page.Value {
+			properties, err := convert.JsonToDict(account.Properties)
+			if err != nil {
+				return err
+			}
+
+			mqlResource, err := CreateResource(runtime, "azure.subscription.cosmosDbService.account",
```
Comment (tas50) [inline providers/azure/resources/cosmosdb.go:137]: Won't be need this in the .lr file?

File: providers/azure/resources/cosmosdb.go [hunk 2]
```diff
@@ -75,10 +96,92 @@ func (a *mqlAzureSubscriptionCosmosDbService) accounts() ([]interface{}, error)
...
```
Comment (preslavgerchev) [inline providers/azure/resources/cosmosdb.go]: i am a bit confused here: the function says its fetching cluster, but we're creating cosmosdb accounts. is this a wrong mql resource here?

File: providers/azure/resources/cosmosdb.go [hunk 3]
```diff
@@ -42,29 +43,49 @@ func (a *mqlAzureSubscriptionCosmosDbServiceAccount) id() (string, error) {
...
+
+	err = fetchPostgreSQLServerGroups(ctx, a.MqlRuntime, conn, subId, &res)
+	if err != nil {
+		return nil, err
+	}
+
+	return res, nil
+}
+
+func fetchCosmosDBAccounts(ctx context.Context, runtime *plugin.Runtime, conn *connection.AzureConnection, subId string, res *[]interface{}) error {
```
Comment (preslavgerchev) [inline providers/azure/resources/cosmosdb.go]: can't we have one function that takes a `type` and filters, based on that? it seems the code is almost identical, except for the fact we're requesting different entities. we can use a param here

Comment (preslavgerchev) [inline providers/azure/resources/cosmosdb.go]:
```suggestion
func fetchCosmosDBAccounts(ctx context.Context, runtime *plugin.Runtime, conn *connection.AzureConnection, subId string) ([]interface{},error) {
```
we should just return the list here, no need to pass a pointer to it to fill it in

File: providers/azure/resources/cosmosdb.go (hunk 2 above)
Comment (HRouhani) [reply providers/azure/resources/cosmosdb.go]: yes usually MondoDB and PostgreSQL are referred as cluster as a logical group, but here in cosmodb context better to stick to the account naming. Updated

File: providers/azure/resources/cosmosdb.go (hunk 3 above)
Comment (HRouhani) [reply providers/azure/resources/cosmosdb.go]: correct, mergged MondoDB and PostgreSQL to one function. Thanks

Comment (HRouhani) [reply providers/azure/resources/cosmosdb.go]: OK, I changed overal to return the list. Thanks

File: providers/azure/resources/cosmosdb.go (hunk 1 above)
Comment (HRouhani) [reply providers/azure/resources/cosmosdb.go:137]: No, already exists, we just append for other types!

File: providers/azure/resources/cosmosdb.go [hunk 4]
```diff
@@ -82,3 +108,46 @@ func (a *mqlAzureSubscriptionCosmosDbService) accounts() ([]interface{}, error)
...
+func fetchMongoDBAndPostgreSQLAccounts(ctx context.Context, runtime *plugin.Runtime, conn *connection.AzureConnection, subId string, resourceType string) ([]interface{}, error) {
+	resClient, err := armresources.NewClient(subId, conn.Token(), &arm.ClientOptions{
+		ClientOptions: conn.ClientOptions(),
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	res := []interface{}{}
+	filter := "resourceType eq '" + resourceType + "'"
```
Comment (preslavgerchev) [inline providers/azure/resources/cosmosdb.go]:
```suggestion
	filter := fmt.Sprintf("resourceType eq '%s'", resourceType)
```

File: providers/azure/resources/cosmosdb.go [hunk 5]
```diff
@@ -82,3 +108,46 @@ func (a *mqlAzureSubscriptionCosmosDbService) accounts() ([]interface{}, error)
 	}
 	return res, nil
 }
+
+func fetchMongoDBAndPostgreSQLAccounts(ctx context.Context, runtime *plugin.Runtime, conn *connection.AzureConnection, subId string, resourceType string) ([]interface{}, error) {
```
Comment (preslavgerchev) [inline providers/azure/resources/cosmosdb.go]: can we rename this func name? to something like `fetchDbAccountsByType`?

File: providers/azure/resources/cosmosdb.go (hunk 4 above)
Comment (HRouhani) [reply providers/azure/resources/cosmosdb.go]: done, good point

File: providers/azure/resources/cosmosdb.go (hunk 5 above)
Comment (HRouhani) [reply providers/azure/resources/cosmosdb.go]: done! no worries

**Comment by preslavgerchev** [comment]
@HRouhani, have you tested if we get back other postgresql and mongodb instances when running this from the `cosmosdb` resource? What if I create a flexible postgresql server, will that end up in the results when querying cosmodb accounts?

**Comment by HRouhani** [comment]
> @HRouhani, have you tested if we get back other postgresql and mongodb instances when running this from the `cosmosdb` resource? What if I create a flexible postgresql server, will that end up in the results when querying cosmodb accounts?

Very good question. No it is not possible:

the PostgreSQL has different resource type which is :  Microsoft.DBforPostgreSQL/flexibleServers

and MondoDB Atlas has different API, need to get it from Market place first.

**Review by preslavgerchev — APPROVED** [review]

**Pull request merged**